src/main/java/jenkins/plugins/git/GitRepoSizeEstimator.java

```diff
@@ -0,0 +1,168 @@
+        boolean useAPI = setSizeFromAPI(remoteName);
+
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        } else {
+            implementation = "DEFAULT";
+        }
+        determineGitTool(implementation);
+    }
+
+    /**
+     * For a given recommended git implementation, validate if the installation exists and provide no suggestion if
+     * implementation doesn't exist.
+     * @param gitImplementation: The recommended git implementation, "git" or "jgit" on the basis of the heuristics.
+     */
+    private void determineGitTool(String gitImplementation) {
+        if (gitImplementation.equals("DEFAULT")) {
+            gitTool = "NONE";
+            return; // Recommend nothing (GitToolRecommendation = NONE)
+        }
+        final Jenkins jenkins = Jenkins.get();
+        GitTool tool = GitUtils.resolveGitTool(gitImplementation, jenkins, null, TaskListener.NULL);
+        if (tool != null) {
+            gitTool = tool.getGitExe();
+        }
+    }
+
+    /**
+     * Determine and estimate the size of a .git cached directory
+     * @param source: Use a {@link AbstractGitSCMSource} to access a cached Jenkins directory, we do not lock it.
+     * @return useCache
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    private boolean setSizeFromCache(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        String cacheEntry = source.getCacheEntry();
+        File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry);
+        if (cacheDir != null) {
+            Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
+            GitClient client = git.getClient();
+            if (!client.hasGitRepo()) {
+                useCache = false;
+            } else {
+                useCache = true;
+                sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
+                sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
+            }
+        } else {
+            useCache = false;
+        }
+        return useCache;
+    }
+
+    /**
+     * Recommend a git implementation on the basis of the given size of a repository
+     * @param sizeOfRepo: Size of a repository (in KiBs)
+     * @return a git implementation, "git" or "jgit"
+     */
+    private String determineSwitchOnSize(Long sizeOfRepo) {
+        if (sizeOfRepo != 0L) {
+            if (sizeOfRepo >= SIZE_TO_SWITCH) {
+                return "git";
+            } else {
+                return "jgit";
+            }
+        }
+        return "NONE";
+    }
+
+    /**
+     * Other plugins can estimate the size of repository using this extension point
+     * The size is assumed to be in KiBs
+     */
+    public static abstract class RepositorySizeAPI implements ExtensionPoint {
+
+        public abstract Long getSizeOfRepository(String remote);
```

Comment:
To clarify what was commented in the meeting
```suggestion
        public abstract Long getSizeOfRepository(String remote);
        
        public abstract boolean acceptsRemote(String remote);
```